Test/tutorial/parser_combinators.dfy

```diff
@@ -0,0 +1,265 @@
+/// # Parser combinators in Dafny
+///
+/// Challenge: define parse combinators.  As a usability test, use them to parse
+/// expressions of the form `(ⁿ)ⁿ` (`n` opening parentheses followed by `n`
+/// closing parentheses).
+///
+/// This challenge is tricky because we can't write what we'd usually write in
+/// Haskell or OCaml:
+///
+/// ```ocaml
+/// let parentheses' () =
+///   (either
+///      (concat (char '(') (concat parentheses' (char ')')))
+///      epsilon)
+/// let parentheses () =
+///   (concat parentheses' eos)
+/// ```
+///
+/// Dafny will reject this definintion definition because thunking isn't enough
+/// to convince it that the `parentheses'` function terminates.
+///
+/// Instead, we need to use stronger specifications to enforce progress, as
+/// shown in the solution below:
+
+module {:options "-functionSyntax:4"} Parsers {
```

Comment:
Since this is a tutorial, maybe we should say something about the `-functionSyntax:4` pragma.